hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java

```diff
@@ -179,25 +193,79 @@ public long takeSnapshot() throws IOException {
     long startTime = Time.monotonicNow();
     TermIndex lastTermIndex = getLastAppliedTermIndex();
     long lastAppliedIndex = lastTermIndex.getIndex();
-    TransactionInfo lastAppliedTrxInfo =
-        TransactionInfo.fromTermIndex(lastTermIndex);
-    if (transactionBuffer.getLatestTrxInfo()
-        .compareTo(lastAppliedTrxInfo) < 0) {
+    if (isInitialized.get()) {
+      TransactionInfo lastAppliedTrxInfo =
+          TransactionInfo.fromTermIndex(lastTermIndex);
+      if (transactionBuffer.getLatestTrxInfo().compareTo(lastAppliedTrxInfo)
+          < 0) {
+        transactionBuffer.updateLatestTrxInfo(
+            TransactionInfo.builder().setCurrentTerm(lastTermIndex.getTerm())
+                .setTransactionIndex(lastTermIndex.getIndex()).build());
+        transactionBuffer.setLatestSnapshot(
+            transactionBuffer.getLatestTrxInfo().toSnapshotInfo());
+      } else {
+        lastAppliedIndex =
+            transactionBuffer.getLatestTrxInfo().getTransactionIndex();
+      }
+
+      transactionBuffer.flush();
+      LOG.info("Current Snapshot Index {}, takeSnapshot took {} ms",
+          lastAppliedIndex, Time.monotonicNow() - startTime);
+    }
+    super.takeSnapshot();
+    return lastAppliedIndex;
+  }
+
+  @Override
+  public void initialize(
+      RaftServer server, RaftGroupId id, RaftStorage raftStorage)
+      throws IOException {
+    super.initialize(server, id, raftStorage);
+    storage.init(raftStorage);
+    loadSnapshot(storage.getLatestSnapshot());
+  }
+
+  private long loadSnapshot(SingleFileSnapshotInfo snapshot)
+      throws IOException {
+    if (snapshot == null) {
```

Comment:
Here load snapshot should be loaded from TransactionInfo table?

Reply:
This is completely removed now. I don't see this fixed.

Reply:
Yes, its not required.

Reply:
for the latest snapshot, getLatestSnapshot is already overriden in SCMStateMachine.

Reply:
But this is not setting lastAppliedIndex, how before loadSnapshot used to set